File: app/src/main/java/com/dumuzeyn/mp3player/UiVisibilityController.java
package com.dumuzeyn.mp3player;

import android.view.View;
import android.view.ViewGroup;

/** Pauses visual work only; playback remains in the foreground Media3 service. */
final class UiVisibilityController {
    private UiVisibilityController() {
    }

    static void apply(MainActivityCore host, boolean visible) {
        if (!visible) {
            host.tabsController.cancelScrollAnimation();
            host.swipeController.cancelForBackground();
            if (host.miniPlayer != null) host.miniPlayer.animate().cancel();
        }
        if (host.particleEffectsView != null) {
            host.particleEffectsView.setUiActive(visible);
        }
        if (host.songsView != null) {
            host.songsView.setHostVisible(visible);
        }
        host.playlistController.setUiActive(visible);
        applyToTree(host.root, visible);
        host.playerUiController.onHostVisibilityChanged(visible);
    }

    private static void applyToTree(View view, boolean visible) {
        if (view == null) return;
        if (view instanceof BackgroundMediaView) {
            ((BackgroundMediaView) view).setUiActive(visible);
        } else if (view instanceof PlayerGradientBackground) {
            ((PlayerGradientBackground) view).setUiActive(visible);
        } else if (view instanceof RotatingCoverImageView) {
            ((RotatingCoverImageView) view).setUiActive(visible);
        } else if (view instanceof SmoothPlaylistTicker) {
            ((SmoothPlaylistTicker) view).setUiActive(visible);
        }
        if (view instanceof ViewGroup) {
            ViewGroup group = (ViewGroup) view;
            for (int index = 0; index < group.getChildCount(); index++) {
                applyToTree(group.getChildAt(index), visible);
            }
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/UiVisibilityController.kt
package com.dumuzeyn.mp3player

import android.view.View
import android.view.ViewGroup

/** Pauses visual work only; playback remains in the foreground Media3 service. */
object UiVisibilityController {
    @JvmStatic
    fun apply(host: MainActivityCore, visible: Boolean) {
        if (!visible) {
            host.tabsController.cancelScrollAnimation()
            host.swipeController.cancelForBackground()
            host.miniPlayer?.animate()?.cancel()
        }
        host.particleEffectsView?.setUiActive(visible)
        host.songsView?.setHostVisible(visible)
        applyToTree(host.root, visible)
        host.playerUiController.onHostVisibilityChanged(visible)
    }

    private fun applyToTree(view: View?, visible: Boolean) {
        if (view == null) return
        when (view) {
            is BackgroundMediaView -> view.setUiActive(visible)
            is PlayerGradientBackground -> view.setUiActive(visible)
            is RotatingCoverImageView -> view.setUiActive(visible)
        }
        if (view is ViewGroup) {
            repeat(view.childCount) { index -> applyToTree(view.getChildAt(index), visible) }
        }
    }
}
